pageserver/src/remote_storage/storage_sync/download.rs

```diff
@@ -33,303 +79,321 @@ pub(super) enum DownloadedTimeline {
     FailedAndRescheduled,
     /// Remote timeline data is found, its latest checkpoint's metadata contents (disk_consistent_lsn) is known.
     /// Initial download successful.
-    Successful,
+    Successful(SyncData<TimelineDownload>),
 }
 
-/// Attempts to download and uncompress files from all remote archives for the timeline given.
+/// Attempts to download all given timeline's layers.
 /// Timeline files that already exist locally are skipped during the download, but the local metadata file is
-/// updated in the end of every checkpoint archive extraction.
+/// updated in the end, if the remote one contains a newer disk_consistent_lsn.
 ///
-/// On an error, bumps the retries count and reschedules the download, with updated archive skip list
-/// (for any new successful archive downloads and extractions).
-pub(super) async fn download_timeline<
-    P: std::fmt::Debug + Send + Sync + 'static,
-    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
->(
-    conf: &'static PageServerConf,
-    remote_assets: Arc<(S, RemoteIndex)>,
```

Comment:
Note that neither download.rs nor upload.rs have an access to `index` now, which removes some common deadlock pitfalls we've experienced recently.